playground/infrastructure/datastore_client.py

```diff
@@ -83,54 +87,67 @@ def save_to_cloud_datastore(self, examples_from_rep: List[Example], sdk: Sdk, or
-                        PrecompiledExample.OUTPUT_EXTENSION.upper(),
-                        PrecompiledExample.LOG_EXTENSION.upper()
-                    ]:
-                        pc_objs_keys_for_removing.append(
-                            self._get_key(DatastoreProps.PRECOMPILED_OBJECT_KIND,
-                                f"{example_id_item}{config.DatastoreProps.KEY_NAME_DELIMITER}{example_type}")
-                        )
-                self._datastore_client.delete_multi(examples_keys_for_removing)
-                self._datastore_client.delete_multi(snippets_keys_for_removing)
-                self._datastore_client.delete_multi(file_keys_for_removing)
-                self._datastore_client.delete_multi(pc_objs_keys_for_removing)
-                logging.info("Finish of deleting extra playground examples ...")
+            snippets.append(snippet)
+
+        if datasets:
+            self._datastore_client.put_multi(datasets)
+        self._datastore_client.put_multi(examples)
+        self._datastore_client.put_multi(snippets)
+        self._datastore_client.put_multi(pc_objects)
+        self._datastore_client.put_multi(files)
+
+        # delete examples from the Cloud Datastore that are not in the repository
+        examples_ids_for_removing = list(filter(lambda key: key not in updated_example_ids, examples_ids_before_updating))
+        if len(examples_ids_for_removing) != 0:
+            logging.info("Start of deleting extra playground examples ...")
+            examples_keys_for_removing = list(
+                map(lambda ex_id: self._get_key(DatastoreProps.EXAMPLE_KIND, ex_id), examples_ids_for_removing)
+            )
+            snippets_keys_for_removing = list(
+                map(lambda ex_id: self._get_key(DatastoreProps.SNIPPET_KIND, ex_id), examples_ids_for_removing)
+            )
+            file_keys_for_removing = list(
+                map(self._get_files_key, examples_ids_for_removing)
+            )
+            pc_objs_keys_for_removing = []
+            for example_id_item in examples_ids_for_removing:
+                for example_type in [
+                    PrecompiledExample.GRAPH_EXTENSION.upper(),
+                    PrecompiledExample.OUTPUT_EXTENSION.upper(),
+                    PrecompiledExample.LOG_EXTENSION.upper()
+                ]:
+                    pc_objs_keys_for_removing.append(
+                        self._get_key(DatastoreProps.PRECOMPILED_OBJECT_KIND, f"{example_id_item}{config.DatastoreProps.KEY_NAME_DELIMITER}{example_type}"))
+            self._datastore_client.delete_multi(examples_keys_for_removing)
+            self._datastore_client.delete_multi(snippets_keys_for_removing)
+            self._datastore_client.delete_multi(file_keys_for_removing)
+            self._datastore_client.delete_multi(pc_objs_keys_for_removing)
+            logging.info("Finish of deleting extra playground examples ...")
```

Comment:
What about datasets removal?

Reply:
I'd create a separate task for that. It looks that we are not able to delete unused datasets here because python scripts are executed for every SDK separately but a dataset can relate to an example from other SDK. Thus, we can compare links that store in GCS and links in _pg_datasets_ kind by a scheduled task in golang to remove them.